gpiozero/mixins.py

```diff
@@ -270,13 +280,119 @@ def _fire_events(self):
+        self._is_held = False
+        self._hold_time = 1
+        self._hold_repeat = False
+        self._hold_thread = RepeatingThread(self)
+
+    def close(self):
+        if self._hold_thread:
+            self._hold_thread.stop()
+            self._hold_thread = None
+        try:
+            super(RepeatingMixin, self).close()
+        except AttributeError:
+            pass
+
+    def _fire_activated(self):
+        super(RepeatingMixin, self)._fire_activated()
+        self._hold_thread.holding.set()
+
+    def _fire_deactivated(self):
+        self._hold_thread.holding.clear()
+        super(RepeatingMixin, self)._fire_deactivated()
+
+    def _fire_held(self):
+        if self.when_held:
+            self.when_held()
+
+    @property
+    def when_held(self):
+        """
+        The function to run when the device has remained active for
+        :attr:`hold_time` seconds.
+
+        This can be set to a function which accepts no (mandatory) parameters,
+        or a Python function which accepts a single mandatory parameter (with
+        as many optional parameters as you like). If the function accepts a
+        single mandatory parameter, the device that activated will be passed
+        as that parameter.
+
+        Set this property to ``None`` (the default) to disable the event.
+        """
+        return self._when_held
+
+    @when_held.setter
+    def when_held(self, value):
+        self._when_held = self._wrap_callback(value)
+
+    @property
+    def hold_time(self):
+        """
+        The length of time (in seconds) to wait after the device is activated,
+        until executing the :attr:`when_held` handler. If :attr:`hold_repeat`
+        is True, this is also the length of time between invocations of
+        :attr:`when_held`.
+        """
+        return self._hold_time
+
+    @hold_time.setter
+    def hold_time(self, value):
+        self._hold_time = value
+
+    @property
+    def hold_repeat(self):
+        """
+        If ``True``, :attr:`when_held` will be executed repeatedly with
+        :attr:`hold_time` seconds between each invocation.
+        """
+        return self._hold_repeat
+
+    @hold_repeat.setter
+    def hold_repeat(self, value):
+        self._hold_repeat = bool(value)
+
+    @property
+    def is_held(self):
+        """
+        When ``True``, the device has been active for at least
+        :attr:`hold_time` seconds.
+        """
+        return self._is_held
+
+
+class RepeatingThread(GPIOThread):
+    """
+    Extends :class:`GPIOThread`. Provides a background thread that repeatedly
+    fires the :attr:`RepeatingEventsMixin.when_held` event as long as the
+    owning device is active.
+    """
+    def __init__(self, parent):
+        super(RepeatingThread, self).__init__(target=self.held, args=(parent,))
+        self.holding = Event()
+        self.start()
+
+    def held(self, parent):
+        while not self.stopping.wait(0):
+            if self.holding.wait(0.1):
```

Comment:
It's a vaguely arbitrary number plucked from the air, but with some background knowledge. The most important rule for background threads is generally: don't suck up all the CPU time. The key to that is to make sure the thread spends the vast majority of its time in an efficient wait state (basically anything like `Event.wait`, `Lock.acquire`, or even `sleep` although the latter isn't that useful).

So, let's look at the code with "n" as our wait there and walk through the various possibilities:

```python
        while not self.stopping.wait(0):
            if self.holding.wait(n):
                while not (
                        self.stopping.wait(0) or
                        parent._inactive_event.wait(parent.hold_time)
                        ):
                    parent._is_held = True
                    parent._fire_held()
                parent._is_held = False
```

Let's say *n* is 0. Now, if the device is inactive `holding` won't be set so we'll be looping around touching only `self.stopping.wait(0)` and `self.holding.wait(0)`. In other words, we're not waiting at all so our thread's going to consume loads of time. Once the device becomes active we start hitting `parent._inactive_event.wait(parent.hold_time)` so assuming the hold time hasn't been set to zero we'll be fine, but as soon as it's inactive again we'll be back to eating CPU.

Let's say *n* is 1. Now our loop's nice and CPU friendly even when the device is inactive. However, what happens when `self.stopping` is set to terminate the loop (when the owning component is closed/destroyed)? Now we've got up to a 1 second wait before the loop will terminate. Okay, not that big of a deal in most scenarios but generally you do want components to shut down in a reasonable length of time (if only because our test suite is likely to go creating and destroying lots of these things, and we don't want to spend too much time hanging around for that!).

So, 0.1 is a vague "guess" I plugged in initially figuring I wanted it to be CPU friendly but not take ages shutting down. I could probably lower that to 0.01 without any serious difference ... perhaps even 0.001. But 0.1's doing shut down quick enough for me at the moment (though I might change my mind once I write some tests for all this!).